### PR TITLE
Rip out non-endomorphism code + dependencies

### PR DESCRIPTION
**Pull request by sipa**
This is a rebased/combined version of the following pull requests/commits with minor changes:
* #825 Switch to our own memcmp function
  * Modification: `secp256k1_memcmp_var` is marked static inline
  * Modification: also replace `memcmp` with `secp256k1_memcmp_var` in exhaustive tests
  * Modification: add reference to GCC bug 95189
* #822 Increase precision of g1 and g2
  * Modification: use the new `secp256k1_memcmp_var` function instead of `memcmp` (see https://github.com/bitcoin-core/secp256k1/pull/822#issuecomment-706610361)
  * Modification: drop the " Allow secp256k1_split_lambda_verify to pass even in the presence of GCC bug https://gcc.gnu.org/bugzilla/show_bug.cgi?id=95189." commit, as it's dealt with using `secp256k1_memcmp_var`.
  * Modification: rename secp256k1_gej_mul_lambda -> secp256k1_ge_mul_lambda
* A new commit that moves the `lambda` constant out of `secp256k1_scalar_split_lambda` and (`_verify`).
* The test commit suggested here: https://github.com/bitcoin-core/secp256k1/pull/822#issuecomment-706610276
  * Modification: use the new accessible `secp256k1_const_lambda` instead of duplicating it.
* #826 Rip out non-endomorphism code
* A new commit that reduces the size of the WNAF output to 129, as we now have proof that the split output is always 128 bits or less.
* A new commit to more consistently use input:`k`, integer outputs:`k1`,`k2`, modulo n outputs:`r1`,`r2`

### CODE REVIEW

**Comment by gmaxwell** [comment]
ACK

**Comment by sipa** [comment]
Comment by @real-or-random, putting it here so I don't forget:

> here’s a nit: the comment for split_lambda is now confusingly far from the actual function. this bit me when looking at it today. the current order is:
> 
> *   huge comment
> *   lambda constant
> *   split_lambda_verify
> *   split_lambda
> 
> we could make it
> 
> *   lambda constant
> *   huge comment
> *   split_lambda
> *   split_lambda_verify
>
> feel free to either fix or ignore

**Comment by gmaxwell** [comment]
putting verify last puts things out of calling order (split_lambda calls _verify). maybe instead just split up the comment and put the range validation proof stuff on the verify part.

File: src/scalar_impl.h [hunk 1]
```diff
@@ -303,21 +318,162 @@ static void secp256k1_scalar_split_lambda(secp256k1_scalar *r1, secp256k1_scalar
...
+ * <    {rounding in c2 and 0 <= k < 2^256}
+ *    2^-1 + 2^256 * |g2/2^384 - (-b1)/d|
+ *  =   {definition of epsilon2}
+ *    2^-1 + epsilon2
+ *
+ * Let
+ *  - k1 = k - c1*a1 - c2*a2
+ *  - k2 = - c1*b1 - c2*b2
+ *
+ * Lemma 3: |k1| < (a1 + a2 + 1)/2 < 2^128
```
Comment (jonasnick) [inline src/scalar_impl.h:372]: Above we say that `k1` has a small size and this lemma only shows that `|k1|` has a small size. Does `k1` refer to different quantities in this text?

Comment (sipa) [reply src/scalar_impl.h:416]: I checked the book (https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.394.3037&rep=rep1&type=pdf), and it says "are small in absolute value". I've updated the comment.

File: src/scalar_impl.h [hunk 2]
```diff
@@ -279,19 +282,31 @@ static void secp256k1_scalar_split_lambda(secp256k1_scalar *r1, secp256k1_scalar
...
  *
- * The algorithm then computes c1 = round(b1 * k / n) and c2 = round(b2 * k / n), and gives
+ * "Guide to Elliptic Curve Cryptography" (Hankerson, Menezes, Vanstone) gives an algorithm
+ * (algorithm 3.74) to find k1 and k2 given k, such that k1 + k2 * lambda == k mod n, and k1
+ * and k2 have a small size.
+ *
+ * The algorithm computes c1 = round(b2 * k / n) and c2 = round((-b1) * k / n), and gives
  * k1 = k - (c1*a1 + c2*a2) and k2 = -(c1*b1 + c2*b2). Instead, we use modular arithmetic, and
- * compute k1 as k - k2 * lambda, avoiding the need for constants a1 and a2.
+ * compute k - k2 * lambda (mod n) which is equivalent to k1 (mod n), avoiding the need for
```
[the remark this answers is not in the document]
Comment (sipa) [reply src/scalar_impl.h]: I've changed this line a bit.

File: src/scalar.h [hunk 3]
```diff
@@ -102,12 +102,11 @@ static void secp256k1_scalar_order_get_num(secp256k1_num *r);
...
-#ifdef USE_ENDOMORPHISM
-/** Find r1 and r2 such that r1+r2*2^128 = a. */
-static void secp256k1_scalar_split_128(secp256k1_scalar *r1, secp256k1_scalar *r2, const secp256k1_scalar *a);
-/** Find r1 and r2 such that r1+r2*lambda = a, and r1 and r2 are maximum 128 bits long (see secp256k1_gej_mul_lambda). */
-static void secp256k1_scalar_split_lambda(secp256k1_scalar *r1, secp256k1_scalar *r2, const secp256k1_scalar *a);
-#endif
+/** Find r1 and r2 such that r1+r2*2^128 = k. */
+static void secp256k1_scalar_split_128(secp256k1_scalar *r1, secp256k1_scalar *r2, const secp256k1_scalar *k);
+/** Find r1 and r2 such that r1+r2*lambda = k,
+ * where r1 and r2 or their negations are maximum 128 bits long (see secp256k1_gej_mul_lambda). */
```
[the remark this answers is not in the document]
Comment (sipa) [reply src/scalar.h]: Good idea, done.

**Comment by luke-jr** [comment]
IMO if GCC bug 95189 is the only problem with normal memcmp, I don't think implementing a new one here is the right solution. At most, check for it in configure and add --no-builtin-memcmp ?

**Comment by sipa** [comment]
@luke-jr I tried writing a test for the bug in autoconfese, but it was more effort than it's worth I think, especially since none of the uses of `memcmp` in libsecp256k1 are performance critical.

**Comment by luke-jr** [comment]
Moving memcmp discussion to #832

**Comment by real-or-random** [comment]
> @luke-jr I tried writing a test for the bug in autoconfese, but it was more effort than it's worth I think, especially since none of the uses of `memcmp` in libsecp256k1 are performance critical. This approach also protects non-autoconf builds.

Agreed, just adding our memcmp is simpler.

**Comment by real-or-random** [comment]
ACK c582abade1c50ef50dc7ee9f7b7af8e06e22065d code inspection, some tests, verified the new g1/g2 constants

File: src/scalar_impl.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (jonasnick) [reply src/scalar_impl.h]: Looks better now.

**Review by jonasnick — COMMENTED** [review]
ACK c582abade1c50ef50dc7ee9f7b7af8e06e22065d

**Pull request merged**